Implement new handling of startup/shutdown/reset reactions in modes

a-sr opened the pull request:
Requires [reactor-c PR #79](https://github.com/lf-lang/reactor-c/pull/79)
Requires [benchmarks PR #15](https://github.com/lf-lang/benchmarks-lingua-franca/pull/15)
Requires [benchmarks PR #18](https://github.com/lf-lang/benchmarks-lingua-franca/pull/18)

Implies [website PR #67](https://github.com/lf-lang/website-lingua-franca/pull/67)
Implies [website PR #68](https://github.com/lf-lang/website-lingua-franca/pull/68)

Closes #999
Closes #1048

---

Implements the following semantics:
- Startup reactions:
    - Only triggered upon first entry into a mode
    - Implication: Some legacy reactors (without reset reactions) might stop working when entered via reset because an "initial token" is not recreated upon reset entry
- Shutdown reactions:
    - Triggered at shutdown time of reactor
    - Only for modes that were active anytime before
    - Executed even is mode is inactive
- New reset reactions:
    - Triggered at activation time of a mode if entered via reset
- New reset transition:
    - No longer resets state variables
        - Can be handled by reset trigger
        - Implication: Always requires manual reset. Legacy reactors (without reset reaction) keep state variables values upon reset, leading to previous "history" leaking into the current (otherwise reset) behavior (e.g. a counter)

---

TODO:
- [x] Fix Python support
- [x] Fix newly blocked keywords in tested models (#947)
- [x] Add a warning for reactor instantiations with state variables but without reset reactions

a-sr commented:
The implementation is complete and this PR and the related ones are ready for review.

The benchmark tests still fail because the fix in the benchmark repository is not yet merged.
The error in cli-tests on Mac looks weird but seems unrelated to this PR, since other tests succeed. @lhstrh Could you help me out with that one?

edwardalee reviewed (approved):
LGTM!  The website documentation will also need to be updated.

lhstrh commented on org.lflang/src/org/lflang/LinguaFranca.xtext:
```diff
@@ -134,7 +134,7 @@ TargetDecl:
             | (braces+='{' (init+=Expression (','  init+=Expression)*)? braces+='}')
         )?
-    ) ';'?
+    ) (reset?='reset')? ';'?
```
I'm confused about this syntax. What does it mean if a state variable of type `foo*` is marked as `reset`?

lhstrh replied:
I also find it a bit strange to see this modifier at the end of the declaration. I think it would make more sense to have it at the beginning.

lhstrh replied:
Also, I'm thinking that a keyword like `transient` might be better, because it is a clear adjective and a much less frequently used term in variable names.

edwardalee replied:
Isn't `reset` already a keyword (for reset transitions)?

edwardalee replied:
I think this needs to have a clear connection with a reset transition.  The state gets reset when a reset transition is taken.

a-sr replied:
The problem with `transient` is that it inverts the modifier and consequently the default (if absent).
Our concept relies on the fact that by default we do not automatically reset a state variable, such that if a reactor was not written with modes in mind we to not leak memory by resetting state variables pointing to allocated memory. The approach with `transient` contradicts this in my opinion.

lhstrh replied:
I see why that's desirable. I still think it's better to have `reset state ...` than `state ... reset`.

Do you know what the reset behavior is for state variables with a pointer types?

a-sr replied:
I added the reset as a shorthand notation for resetting the state variable in a reset reaction, so I agree with @edwardalee that it should have a connection to it.

a-sr replied:
In my mind, I do not consider the reset as an integral component of the state variable but rather a handler added to it, like a deadline or an after delay. I also did not want to draw to much attention to it by putting it first but if you think a `reset state` makes more sense, I can change that.

> Do you know what the reset behavior is for state variables with a pointer types?

Well, the variable is reset to the initial value provided by the user. I assumed that there always has to be one but if that was a misconception, I will add an error message in case `reset` is used without an initial value.

lhstrh replied:
Handlers go at the end, sure, but this isn't really a handler but a modifier that labels the state as being automatically reset. I suggest we put the `reset` in front of the state variable declaration.

a-sr replied:
Ok

lhstrh commented on org.lflang.ui/src/org/lflang/ui/highlighting/LFSemanticHighlightingCalculator.java:
```diff
@@ -61,23 +56,6 @@ public void provideHighlightingFor(XtextResource resource, IHighlightedPositionA
         }
         Model model = (Model) parseResult.getRootASTElement();
-        
-        // Provide keyword highlighting for special mode transitions
```
How come this is no longer necessary?

a-sr replied:
When there was only `reset(<mode>)` and `continue(<mode>)`, I was able to make a workaround that treated these keywords as strings in the grammar (#947) and then faked their role as keywords in the editor by adjusting the highlighting (which also only worked in Epoch). However, with `reset` now used in the enum rule for built-in triggers and as boolean modifier (reset?='reset') in state variables, it was not possible to treat is as a non-keyword. So I also rolled back the workaround in transition effect to make the grammar more consistent.

lhstrh reviewed (changes requested):
This looks good to me, except for the syntax (as pointed out in conversation). If also really like your suggestion to validate that any state variable that is marked as reset indeed has an initial state assigned to it.

a-sr commented:
Ok
- added the validation error for reset state variables without initial values,
- moved the `reset` modifier in front of the `state` keyword, and
- adjusted the syntax examples on the website ([PR68](https://github.com/lf-lang/website-lingua-franca/pull/68))

a-sr commented:
Rust benchmark tests currently fail because there are newly added models using `reset` as variable name. [Benchmarks PR #18](https://github.com/lf-lang/benchmarks-lingua-franca/pull/18) fixes these models.

edwardalee commented:
Is this ready to merge?

lhstrh commented on org.lflang/src/org/lflang/validation/LFValidator.java:
```diff
@@ -1313,6 +1308,79 @@ public void checkModeInstanceNamespace(Reactor reactor) {
     }
 
+    @Check(CheckType.FAST)
+    public void checkMissingStateResetInMode(Reactor reactor) {
```
We should probably add some unit tests for this...

a-sr replied:
Added unit tests for all mode related validation rules.

lhstrh commented on test/C/src/modal_models/ModalStateResetAuto.lf:
```diff
@@ -0,0 +1,98 @@
+    modal.count1,
+    modal.count2
+    -> test.events
+}
```
Missing newline.

a-sr replied:
Added newline.

lhstrh reviewed (commented):
These changes are looking great! My only comment is that the new validator checks should have unit tests. Other than that, this looks ready to merge.

lhstrh commented on test/C/src/modal_models/ModalStateResetAuto.lf:
```diff
@@ -0,0 +1,98 @@
+    =}
+    
+    initial mode One {
+        state counter1:int(0);
```
It looks like the state variable isn't rendered quite correctly in the diagram:
![image](https://user-images.githubusercontent.com/19938940/173119911-40db4248-d00a-4c1a-85be-d855e5785d9d.png)

lhstrh commented on org.lflang/src/org/lflang/validation/LFValidator.java:
```diff
@@ -1304,15 +1299,88 @@ public void checkModeInstanceNamespace(Reactor reactor) {
     }
 
+    @Check(CheckType.FAST)
+    public void checkMissingStateResetInMode(Reactor reactor) {
```
It appears that a `reset` transition is the default, meaning that is no modifier is given in the transition effect, this is to mean that it is a reset transition. This could be confusing, because it may suggest that a "normal" transition is of a different kind than `reset` or `continue`. We had a similar problem with actions, and decided that it would be better to require a disambiguation from the programmer and thus only allow `logical action` or `physical action`, and not just `action`. I think we should take the same approach here, and enforce this in the validator so we can present a useful error message.

lhstrh commented on org.lflang/src/org/lflang/LinguaFranca.xtext:
```diff
@@ -451,6 +453,12 @@ enum ActionOrigin:
+    STARTUP='startup' | SHUTDOWN='shutdown' | RESET='reset';
+
+enum ModeTransition:
+    RESET='reset' | HISTORY='continue';
```
I'm still a bit on the fence about the `continue` keyword. If refer to this a "history transition" and label it with an "H" in the diagram, then why don't we just use the `history` keyword and avoid all confusion?

lhstrh commented on test/Python/src/modal_models/ModalStateReset.lf:
```diff
@@ -29,6 +29,10 @@ reactor Modal {
         state counter1(0)
         timer T1(0msec, 250msec)
 
+        reaction(reset) {=
```
The rendering of the `reset` trigger seems a bit off:

![image](https://user-images.githubusercontent.com/19938940/173121919-d5c4b2b9-ace4-4295-ba36-688b01801f62.png)

I would actually like to propose an alternative visual syntax. How about something along these lines:
![image](https://user-images.githubusercontent.com/19938940/173123030-6fa9810b-51de-403c-a57e-f133314b2460.png)

lhstrh commented on org.lflang.tests/src/org/lflang/tests/compiler/LinguaFrancaValidationTest.java:
```diff
@@ -2221,7 +2222,153 @@ public void testUnrecognizedTarget() throws Exception {
 
-    
+    @Test
+    public void testInitialMode() throws Exception {
```
Great! Thanks for putting these in!

lhstrh reviewed (approved):
This looks really good. I raised a couple points about syntax, validation, and visual representation. If you have the bandwidth, I would prefer we address these issues as part of this PR... I'm also fine with merging and adding fixes later.

lhstrh commented:
I'll go ahead and merge this in to avoid stalling things any further. I've created issues to track my comments.
